File: src/main/java/com/bazbatlabs/smashballs/controllers/WorldController.java
package com.bazbatlabs.smashballs.controllers;

import javax.microedition.khronos.opengles.GL10;
import android.view.KeyEvent;

import tv.ouya.console.api.OuyaController;

import com.bazbatlabs.smashballs.models.*;
import com.bazbatlabs.smashballs.views.WorldView;

public final class WorldController implements Controller {

    private final World world;
    private final WorldView view;
    private final Artist artist;

    public WorldController(Artist artist, ImageMap images, SoundMap sounds) {
        WorldEvents events = new WorldEvents();

        this.world = new World(events);
        this.view = new WorldView(this.world, events, images, sounds, artist);
        this.artist = artist;
    }

    @Override
    public void draw(int screenWidth, int screenHeight) {
        view.draw();
    }

    @Override
    public Controller update() {
        world.update();
        return this;
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        boolean handled = true;

        switch (keyCode) {
        case OuyaController.BUTTON_DPAD_LEFT:
            world.startAcceleratingPaddle(Direction.LEFT);
            break;

        case OuyaController.BUTTON_DPAD_RIGHT:
            world.startAcceleratingPaddle(Direction.RIGHT);
            break;

        default:
            handled = false;
            break;
        }

        return handled;
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        boolean handled = true;

        switch (keyCode) {
        case OuyaController.BUTTON_DPAD_LEFT:
            world.stopAcceleratingPaddle(Direction.LEFT);
            break;

        case OuyaController.BUTTON_DPAD_RIGHT:
            world.stopAcceleratingPaddle(Direction.RIGHT);
            break;

        case OuyaController.BUTTON_O:
            world.kickstartBall();
            break;

        default:
            handled = false;
            break;
        }

        return handled;
    }

    @Override
    public void changeSurface(int width, int height) {
        artist.changeSurface(width, height);
    }
}

File: src/main/scala/controllers/WorldController.scala
package com.bazbatlabs.bricksmash.controllers

import scala.io.Source

import android.view.KeyEvent

import tv.ouya.console.api.OuyaController
import android.content.res.Resources

import com.bazbatlabs.bricksmash.lib._
import com.bazbatlabs.bricksmash.models._
import com.bazbatlabs.bricksmash.views._
import com.bazbatlabs.bricksmash.views.WorldView
import com.bazbatlabs.bricksmash.R

class WorldController(brickLayout: Array[Array[Option[Brick.Type.Value]]], val level: Int,
                      val artist: Artist, val images: ImageMap,
                      val sounds: SoundMap, resources: Resources) extends Controller {

  val events = new WorldEvents()
  val world = new World(brickLayout, events)
  val view = new WorldView(world, events, images, sounds, artist)

  override def draw(screenWidth: Int, screenHeight: Int) { view.draw() }

  override def update(): Controller = {
    world.update()

    if (world.isReadyToMoveToNextLevel) {
      new LevelLoaderController(level + 1, artist, images, sounds, resources)
    } else if (world.isReadyToResetToMenu) {
      new MenuController(artist, images, sounds, resources)
    } else {
      this
    }
  }

  override def onKeyDown(keyCode: Int, event: KeyEvent): Boolean = {
    var handled = true

    keyCode match {
      case OuyaController.BUTTON_DPAD_LEFT => world.startAcceleratingPaddle(Direction.Left)
      case OuyaController.BUTTON_DPAD_RIGHT => world.startAcceleratingPaddle(Direction.Right)
      case _ => handled = false
    }

    handled
  }

  override def onKeyUp(keyCode: Int, event: KeyEvent): Boolean = {
    var handled = true

    keyCode match {
      case OuyaController.BUTTON_DPAD_LEFT => world.stopAcceleratingPaddle(Direction.Left)
      case OuyaController.BUTTON_DPAD_RIGHT => world.stopAcceleratingPaddle(Direction.Right)
      case OuyaController.BUTTON_O => world.kickstartBall()
      case _ => handled = false
    }

    handled
  }

  override def changeSurface(width: Int, height: Int) {
    artist.changeSurface(width, height)
  }
}
